JetMETCorrections/Type1MET/interface/JetCleanerForType1METT.h

```diff
@@ -0,0 +1,222 @@
+    // (because pat::Jet->p4() returns the JES corrected, not the raw, jet momentum)
+    // But it does not handle the muon removal!!!!! MM
+    {
+    public:
+
+      RawJetExtractorT(){}
+      reco::Candidate::LorentzVector operator()(const T& jet) const
+	{
+	  return jet.p4();
+	}
+    };
+}
+
+
+template <typename T, typename Textractor>
+class JetCleanerForType1METT : public edm::stream::EDProducer<>
+{
+ public:
+
+  explicit JetCleanerForType1METT(const edm::ParameterSet& cfg):
+    moduleLabel_(cfg.getParameter<std::string>("@module_label")),
+    offsetCorrLabel_(""),
+    skipMuonSelection_(nullptr)
+      {
+       token_ = consumes<std::vector<T> >(cfg.getParameter<edm::InputTag>("src"));
+
+       offsetCorrLabel_ = cfg.getParameter<edm::InputTag>("offsetCorrLabel");
+       offsetCorrToken_ = consumes<reco::JetCorrector>(offsetCorrLabel_);
+
+       jetCorrLabel_ = cfg.getParameter<edm::InputTag>("jetCorrLabel"); //for MC
+       jetCorrLabelRes_ = cfg.getParameter<edm::InputTag>("jetCorrLabelRes"); //for data
+       jetCorrToken_ = mayConsume<reco::JetCorrector>(jetCorrLabel_);
+       jetCorrResToken_ = mayConsume<reco::JetCorrector>(jetCorrLabelRes_);
+       
+       jetCorrEtaMax_ = cfg.getUntrackedParameter<double>("jetCorrEtaMax",9.9);
+       
+       type1JetPtThreshold_ = cfg.getParameter<double>("type1JetPtThreshold");
+
+       skipEM_ = cfg.getParameter<bool>("skipEM");
+       if ( skipEM_ ) {
+           skipEMfractionThreshold_ = cfg.getParameter<double>("skipEMfractionThreshold");
+       }
+       
+       skipMuons_ = cfg.getParameter<bool>("skipMuons");
+       if ( skipMuons_ ) {
+           std::string skipMuonSelection_string = cfg.getParameter<std::string>("skipMuonSelection");
+           skipMuonSelection_.reset( new StringCutObjectSelector<reco::Candidate>(skipMuonSelection_string,true) ) ;
+       }
+
+       produces<std::vector<T> >();
+      }
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+        edm::ParameterSetDescription desc;
+        desc.add<std::string>("@module_label");
+        desc.add<edm::InputTag>("src");
+        desc.add<edm::InputTag>("offsetCorrLabel");
+        desc.add<edm::InputTag>("jetCorrLabel");
+        desc.add<edm::InputTag>("jetCorrLabelRes");
+        desc.addUntracked<double>("jetCorrEtaMax",9.9);
+        desc.add<double>("type1JetPtThreshold");
+        desc.add<bool>("skipEM");
+        desc.add<double>("skipEMfractionThreshold");
+        desc.add<bool>("skipMuons");
+        desc.add<std::string>("skipMuonSelection");
+        descriptions.addDefault(desc);
+    }
+
+ private:
+
+  void produce(edm::Event& evt, const edm::EventSetup& es)
+  {
+
+    edm::Handle<reco::JetCorrector> jetCorr;
+    //automatic switch for residual corrections
+    if(evt.isRealData() ) {
+      jetCorrLabel_ = jetCorrLabelRes_;
+      evt.getByToken(jetCorrResToken_, jetCorr);
```

Comment:
... aside: we requested at some point from JME/JEC to provide dummy residual corrections for MC so that the same code can be used for both.
It would help to check on this again.